### ISSUE TITLE
Questions about the `TRF().train()` method

### ISSUE THREAD

**Issue by xinyiguan** [powerfulbean/mTRFpy#30]
Hi mTRFpy developers, 

I have some questions about `TRF().train()` method. 

1. The k value in the `TRF().train()` function. In the comment, it writes that the default is set to 5, but I see that in the function argument, it is set to `k=-1`. So just to confirm, the default is the leave-one-out cross validation, right?

2. In the original MATLAB implementation, we are able to partition data into separate training and test sets with `mTRFpartition` function. Is there an equivalent function implemented in this library? Or that we need to manually split the training and test set first, feed the training set to the `TRF().train()` to train and cross-eval the model, and then do `TRF().predict()` with the test set?

**Comment by OleBialas** [comment]
Hi, thanks for catching that! You are correct, the documentation is wrong and leave-one-out cross-validation is the default.
mTRFpy does not implement the functionality of mTRFpartition, you can do that with [numpy.array_split](https://numpy.org/doc/stable/reference/generated/numpy.array_split.html).
Assuming you want a random 10% of your data for testing, you could do something like this:

```python
import numpy as np
from mtrf import load_sample_data

n = 10 # number of segments
stim, resp, fs = load_sample_data()
stim, resp = np.array_split(stim, n), np.array_split(resp, n)  # split the data
idx = np.random.choice(n, n, replace=False)  # randomize segment order
resp_train, resp_test = [resp[i] for i in idx[:n-1]], [resp[-1]] 
stim_train, stim_test = [stim[i] for i in idx[:n-1]], [stim[-1]]
```

You may also be interested in the [nested_crossval](https://mtrfpy.readthedocs.io/en/latest/api.html#mtrf.stats.nested_crossval) function. Rather than picking an arbitrary test segment, this allows you to rotate the test data and average across all segments

**Comment by xinyiguan** [comment]
Hi @OleBialas 

Thanks for response and the tips! I have a few follow up questions and hopefully I can get some more clarifications from you. 

I am trying to compare the mTRF analysis using MATLAB and the python. In MATLAB, we will do something like below for the model training and testing procedure:

```matlab
% 1. split dataset in training and testing (create 10 folds: 9 training, 1 test, and specify fold 5 as test set)
[sTrain,rTrain,sTest,rTest] = mTRFpartition(stim,resp,10,5);

%  2. conduct a cross-validation procedure on the training set to identify the optimal value for lambda
cv = mTRFcrossval(sTrain,rTrain,fs,Dir,tmin,tmax,lambda);

% 3. average the performance metrics across folds, and take the 𝜆 corresponding to the maximum correlation value.
[rmax,idx] = max(mean(cv.r));

% 4. use this lambda value to train our final model 
model = mTRFtrain(sTrain,rTrain,fs,Dir,tmin,tmax,lambda(idx));

% 5. test the model on the held-out test set
[pred,stats] = mTRFpredict(sTest,rTest,model);

```

To do a similar thing in python with this package, and continue on your example code above, we will do something like the following, correct? :

```python3
from mtrf.model import TRF
from mtrf.stats import crossval

tfr = TRF(direction=1)
trf.train(stimulus= stim_train, response= resp_train, fs=64, tmin=0, tmax=0.4, regularization=1000)

# test and predict:
prediction, r_fwd = trf.predict(stimulus, response)

# conduct cross evaluation procedure on the training set to identify the optimal lambda
r_fwd = crossval(model=trf, stimulus= stim_test, response= resp_test, fs=64, tmin=0, tmax=0.4, regularization=1000)

```

Questions here: 

1. to my understanding, the `TRF().train()` function covers the MATLAB steps 2 to 4, correct? The `train()` function already include the cross-validation steps to select the best lambda.
2. I am not sure what should go in the `crossval` function. Should I feed the function with the test set (stim_test, and resp_test) or the entire stim and resp (stim, resp)?
3. Similarly, what is the correct way to use the `nested_crossval`function? What should the inputs be for the arguments `stimulus` and `response`? I guess in general, I am not sure about how the cross-validation steps fit in the procedure here. 

btw, you might also want to amend the comments of the default k value in the `nested_crossval` function as well. 

Thanks in advance for any clarifications you can provide!

**Comment by OleBialas** [comment]
The `TRF.train` method uses cross-validation to find the best lambda if you pass a list of candidate values for the regularization parameter. If, like in your example, you are using a single regularization value, it will simply fit the TRF to the data using that regularization value.

In your scenario, there is no need for cross-validation. You can simply train and test on the respective parts of your data.
Cross-validation is necessary if you want to rotate your test and train set.

The `nested_crossval` function deals with the problem of overfitting that results from optimizing lambda for a given data set, this is [explained here](https://mtrfpy.readthedocs.io/en/latest/optimization.html#more-overfitting)

hope that helps!
